TICL reconstruction update

rovere opened the pull request:
That's a backport of #31906 

It's not exactly a _perfect_ backport, but it introduces the same functionalities and improvements, together with the needed components.

Please refer to that PR to see what has been changed and what to expect from this PR, too.

mariadalfonso committed:
Backport of #29837, w/o Configuration/StandardSequences changes

apsallid committed:
Backport of #31730

felicepantaleo committed:
Backport #30007

gartung committed:
Backport of #30815

JamminJones committed:
Partial backport of #31013

Dr15Jones committed:
Backport #31031

Dereferencing an uninitialized edm::Handle leads to dereference
of a nullptr which is undefined behavior. Rearranged code to
avoid using edm::Handles at all.

felicepantaleo committed:
Created a ClusterFilter by Algo, Size and Layer Range

felicepantaleo committed:
include an Electromagnetic TrackSeeded iteration as first iteration

felicepantaleo committed:
Add the range of layers to the filter parameters

felicepantaleo committed:
Inject tracksters from the track seeded electromagnetic interation in the merged collection

felicepantaleo committed:
create empty multiclusters

rovere committed:
Add TrkEM tracksters into the Merge collection

rovere committed:
Select trackster-PID also on EM/Total trackster energy ratio

rovere committed:
Fix merge conflict.

felicepantaleo committed:
Fix input collection for TRK iteration to be the output of EM iteration

felicepantaleo committed:
Sorting seeding regions by track momentum

felicepantaleo committed:
Use only clusters with minimum size of 3

felicepantaleo committed:
Filter only with a maxLayerId in TRKEM iteration.

felicepantaleo committed:
Remove MIP iteration

felicepantaleo committed:
clang format

felicepantaleo committed:
Limit Global EM iteration to layer 30

felicepantaleo committed:
Limit track seeded iterations to tracks with at most 2 missing outer hits in the Tracker

felicepantaleo committed:
Add TrkEM products to EventContent

felicepantaleo committed:
Remove missing layers cut for EM based iterations

felicepantaleo committed:
Introducing a shower_start_max_layer cut

felicepantaleo committed:
raising shower_start_max_layer from 3 to 7 in EM iterations

felicepantaleo committed:
Count usage of layer clusters, only for those tracksters passing the selection

felicepantaleo committed:
Increasing missing outer hits from 3 to 5

felicepantaleo committed:
increase missing layer from 0 to 1 and loosen min_cos_theta to 0.961 in EM iterations

felicepantaleo committed:
tighten min_cos_theta from 0.961 to 0.978 in EM iterations

felicepantaleo committed:
Rename parameter missing_layers to skip_layers.

felicepantaleo committed:
Retuning of PR cuts in EM iterations

felicepantaleo committed:
Introducing two cuts: a max_missing_layers_in_trackster and a minimum number of different layers in a trackster

felicepantaleo committed:
Refactoring

rovere committed:
Update cuts

felicepantaleo committed:
Make max_missing_layers_in_trackster cut inclusive

felicepantaleo committed:
code format

felicepantaleo committed:
Move trackster energy calculation after the selections are applied

felicepantaleo committed:
Add a new max_longitudinal_sigmaPCA cut for EM iterations, with an initial value of 10

felicepantaleo committed:
Add a new cut: root_doublet_max_distance_from_seed for TrkEM iteration, initial value 0.05

felicepantaleo committed:
clang format

felicepantaleo committed:
fix trackster to seed association

felicepantaleo committed:
Setting  max_missing_layers_in_trackster to 2 from 0 in EM iterations

rovere committed:
Fix compilation error in C++

rovere committed:
Remove duplicate python parameter

felicepantaleo committed:
Sorting seeding regions by pT rather than p

felicepantaleo committed:
Added TICL PF interpretation.

felicepantaleo committed:
Feed PFTICL the correct InputTag

felicepantaleo committed:
Adapt event content

rovere committed:
Protect against empty multiclusters (was 7f2d245ca24)

rovere committed:
Fix hitpattern call

rovere committed:
Code format

cmsbuild commented:
A new Pull Request was created by  @rovere (Marco Rovere) for CMSSW_11_1_X.

It involves the following packages:

DataFormats/HGCalReco
RecoEgamma/EgammaTools
RecoHGCal/Configuration
RecoHGCal/TICL
RecoLocalCalo/HGCalRecAlgos
RecoLocalCalo/HGCalRecProducers
RecoParticleFlow/PFClusterProducer
Validation/HGCalValidation


@perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please review it and eventually sign? Thanks.
@felicepantaleo, @jainshilpi, @argiro, @bsunanda, @pfs, @thomreis, @varuns23, @seemasharmafnal, @mmarionncern, @sethzenz, @lgray, @apsallid, @vandreev11, @Sam-Harper, @cbernet, @cseez, @deguio, @hatakeyamak, @clelange, @rchatter, @riga, @edjtscott, @sobhatta, @lecriste, @afiqaize this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

rovere commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: aa350cca6dd31e23acbb2419d8d614f16694d1f3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10219/summary.html
CMSSW: CMSSW_11_1_X_2020-10-22-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10219/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2552 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2784776
 - DQMHistoTests: Total failures: 6002
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778724
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

jfernan2 commented:
@rovere I understand we should not expect the same changes as in the master #31906 since they are not identical, indeed the number of changes is quite different.
However in both cases plots in HGCAL / TICLPFCandidates / 3  get now zero entries and HGCAL / HGCalValidator / ticlMultiClustersFromTrackstersMerge/Trk now get many Denominator null values
https://tinyurl.com/y3zx7cww
https://tinyurl.com/yyrcgaoe

I guess this is just a matter of a new definition, but I wonder if it is expected. I will post this same question in master PR

felicepantaleo committed:
Review comments

felicepantaleo committed:
Provide EM and HAD fraction in HGCAL for JetMET for calibration

In the process of implementing the em and had energy
fractions for the PFCandidates produced inside HGCal, we
also simplified their energy assignment.

rovere committed:
Backport of #31881

rovere committed:
Manual port of f4c1e2f1f7b and 55110b5d259

rovere committed:
Code format

cmsbuild commented:
Pull request #31907 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

rovere commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ed00efd4d84e1ade61740d389ff8bb6ab73e1043
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10420/summary.html
CMSSW: CMSSW_11_1_X_2020-10-30-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10420/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2554 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2784776
 - DQMHistoTests: Total failures: 6003
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778723
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

felicepantaleo committed:
assign track momentum direction at point of closest approach instead of trackster's barycenter in case of charged particles

felicepantaleo committed:
Additional review comments

jfernan2 commented:
With this PR plots in HGCAL / TICLPFCandidates / 3 get now zero entries and HGCAL / HGCalValidator / ticlMultiClustersFromTrackstersMerge/Trk now get many Denominator null values
Is this expected understood?
Thanks

cmsbuild commented:
Pull request #31907 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

rovere commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

rovere commented:
> With this PR plots in HGCAL / TICLPFCandidates / 3 get now zero entries and HGCAL / HGCalValidator / ticlMultiClustersFromTrackstersMerge/Trk now get many Denominator null values
> Is this expected understood?
> Thanks

Ciao @jfernan2 
thanks for the detailed checks and report.
Category 3 in PFCandidates should not be a problem and is known.
I'm not sure I was able to spot the histograms that had the null values: would you mind being a little more specific?

cmsbuild commented:
+1
Tested at: dbe2815976eea27548700b18f3f5ca587e53531c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10432/summary.html
CMSSW: CMSSW_11_1_X_2020-11-01-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10432/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2554 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2784776
 - DQMHistoTests: Total failures: 6004
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778722
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

jfernan2 commented:
Thanks @rovere 
If category 3 candidates are not a problem, then it is fine, because the second part of my last comment has disappeared with last commit from @felicepantaleo 
In previous tests there were histograms with large number of entris in the bin 0 for Denom* MEs.
Thanks

jfernan2 commented:
+1

apsallid committed:
Backport of #32021

felicepantaleo committed:
Make track cuts configurable

Commit:
Backport of #32022

Restore distance cut in CLUE position calculation

perrotta commented:
@rovere please let us know when this PR will get sinchronized with the one recently merged in the master, so that a review of this backport PR can start

felicepantaleo committed:
Remove unused code

cmsbuild commented:
Pull request #31907 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

rovere commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 48bcd3185655b65300c2ceafbc07a262fd0c723c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10761/summary.html
CMSSW: CMSSW_11_1_X_2020-11-15-0000
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10761/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3937 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2784828
 - DQMHistoTests: Total failures: 10746
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2774032
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

rovere commented:
> @rovere please let us know when this PR will get sinchronized with the one recently merged in the master, so that a review of this backport PR can start

Ciao @perrotta at present this PR includes all the code-changes we would like to propose for inclusion in `CMSSW_11_1`. This has several backports (needed) included, but in itself, it has very few code-changes with respect to the corresponding code in `CMSSW_11_2`. The only changes are due to evolution in the interface of a couple of objects that we decided it was best not to backport also in this branch. The most important one is the query on the number of missing outer hits for a track, that went through a substantial improvement in `11_2` that is still missing here.
Let us know if you have specific questions. The bulk of the discussion should have happened already while integrating #31906

jfernan2 commented:
@rovere I am quite confused by this PR and the changes involved. I understand it is not a exact backport of #31906  but there are MEs modified in Muons, BTAG, Tracking, Taus, JetMET,... and it is not in every 2026 workflow
In some cases it looks just like a precision issue and in other the PR brings statistics to previously empty plots but I wonder if somehow this is understood.
Thanks

silviodonato commented:
@jfernan2 could you point us some examples? It looks like all changes are related to electrons, as expected

jfernan2 commented:
Yes @silviodonato  Everything is here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2020-11-15-0000+f9b338/39903/dqm-histo-comparison-summary.html

In particular:
https://tinyurl.com/y59ft6ow
https://tinyurl.com/y3qltzup
https://tinyurl.com/y45fr4t4
https://tinyurl.com/y2s92bbp
https://tinyurl.com/y2zsnqj4
https://tinyurl.com/y6j88sg3
https://tinyurl.com/y4ll6t25

rovere commented:
Dear all,
@jfernan2 is correct! And thanks a lot for reporting this!

* The results of the testing of the full backport (i.e. this PR as is at 48bcd31)  is the following:                                                                                                                                                      
   - 20034.0 baseline GUI (1566) 20034.0 pr GUI (1566) 20034.0 overlay GUI (1566)
   - 20434.0 baseline GUI (2774) 20434.0 pr GUI (2774) 20434.0 overlay GUI (2774)
   - 21234.0 baseline GUI (1639) 21234.0 pr GUI (1639) 21234.0 overlay GUI (1639)
   - 23234.0 baseline GUI (2996) 23234.0 pr GUI (2996) 23234.0 overlay GUI (2996)
   - This can be translated in terms of _regression-failing_ geometries:
   - 20034 ==> D35/C4/v9  ==> **PASS**
   - 20434 ==> D41/C8/v10 ==> **FAIL**
   - 21234 ==> D44/C6/v9  ==> **PASS**
   - 23234 ==> D49/C9/v11 ==> **FAIL** (HLT TDR)

   - So, geometry `v9` is not affected by the _unexpected regression_ problem.
   - `v9` uses the following customization era: `phase2_hgcalV9`
   - in this era, there is no mention of the scintillator correction factor (i.e. of its regional em factor) : it remains at its default value of 1.
   - Therefore, the changes introduced by the backports #31730 and #32021 of the regional em factors, and their usage in the `RecHit` reconstruction do not affect `v9` geometry. This will silently keep on working with the old factor of 1.
   - The other geometries, instead, will see the improved regional em factors **applied directly are `RecHit` level**
   - The way in which this change will affect the rest of the reconstruction in apparently _unexpected_ way is via the `PFRecHitProducer` that will consume the `HGCAL` `RecHits`. Since those changed in the scintillator part, the rest of the reconstruction that will consume the `PFRechHitProducer:particleFlowRecHitHGC` will be affected as well. And this is what we see.
   - The original #31906 **did not have the regional em factors active** during the test, therefore the changes were limited to the _expected_ ones.
   - In fact, it would have been nice to have the results of the tests from #31730 and #32021 still available, to confirm this even further, but unfortunately, they are no longer available.

jfernan2 commented:
Thanks for the explanation @rovere 
The Jenkins test of #32021 are still available and showed no differences. I wonder if the Jenkins test could be relaunched for #31730 despite it is a closed PR, I did not inspect that PR by that time since it did not involve changes in DQM packages

rovere commented:
> Thanks for the explanation @rovere
> The Jenkins test of #32021 are still available and showed no differences. I wonder if the Jenkins test could be relaunched for #31730 despite it is a closed PR, I did not inspect that PR by that time since it did not involve changes in DQM packages

Ciao @jfernan2 in fact #32021 is more a bug fix, while the real changes were introduced by the other one, #31730 
I'll let others comment on the other request you made.

silviodonato commented:
@jfernan2 @rovere I made https://github.com/cms-sw/cmssw/pull/32165 to help with the review. The PR is a copy of #31907 after having removed #32021 and #31730

rovere commented:
Just for my education: is there a way to tell the bot to test a specific PR at a precise commit?

silviodonato commented:
From https://github.com/cms-sw/cmssw/pull/32165#issuecomment-729195748 . 

> 
> Comparison is ready
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d355e5/10809/summary.html
> 
> Comparison Summary:
>  - No significant changes to the logs found
>  - Reco comparison results: 2552 differences found in the comparisons
>  - DQMHistoTests: Total files compared: 36
>  - DQMHistoTests: Total histograms compared: 2784828
>  - DQMHistoTests: Total failures: 6002
>  - DQMHistoTests: Total nulls: 0
>  - DQMHistoTests: Total successes: 2778776
>  - DQMHistoTests: Total skipped: 50
>  - DQMHistoTests: Total Missing objects: 0
>  - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
>  - Checked 152 log files, 16 edm output root files, 36 DQM output files
> 
> 

As expected, there are changes in variables related to TICL. I see also some changes in EgammaV/ElectronMcFakeValidator, ElectronMcSignalValidator, ElectronMcSignalValidatorPt1000. @rovere did you expect that?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2020-11-17-1100+d355e5/39938/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44/EgammaV.html

rovere commented:
> From [#32165 (comment)](https://github.com/cms-sw/cmssw/pull/32165#issuecomment-729195748) .
> 
> > Comparison is ready
> > https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d355e5/10809/summary.html
> > Comparison Summary:
> > 
> > * No significant changes to the logs found
> > * Reco comparison results: 2552 differences found in the comparisons
> > * DQMHistoTests: Total files compared: 36
> > * DQMHistoTests: Total histograms compared: 2784828
> > * DQMHistoTests: Total failures: 6002
> > * DQMHistoTests: Total nulls: 0
> > * DQMHistoTests: Total successes: 2778776
> > * DQMHistoTests: Total skipped: 50
> > * DQMHistoTests: Total Missing objects: 0
> > * DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
> > * Checked 152 log files, 16 edm output root files, 36 DQM output files
> 
> As expected, there are changes in variables related to TICL. I see also some changes in EgammaV/ElectronMcFakeValidator, ElectronMcSignalValidator, ElectronMcSignalValidatorPt1000. @rovere did you expect that?
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2020-11-17-1100+d355e5/39938/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44/EgammaV.html

Yes, I do expect them since  #31527 by @SohamBhattacharya is feeding electrons with `TICL` products.

SohamBhattacharya commented:
> > From [#32165 (comment)](https://github.com/cms-sw/cmssw/pull/32165#issuecomment-729195748) .
> > > Comparison is ready
> > > https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d355e5/10809/summary.html
> > > Comparison Summary:
> > > 
> > > * No significant changes to the logs found
> > > * Reco comparison results: 2552 differences found in the comparisons
> > > * DQMHistoTests: Total files compared: 36
> > > * DQMHistoTests: Total histograms compared: 2784828
> > > * DQMHistoTests: Total failures: 6002
> > > * DQMHistoTests: Total nulls: 0
> > > * DQMHistoTests: Total successes: 2778776
> > > * DQMHistoTests: Total skipped: 50
> > > * DQMHistoTests: Total Missing objects: 0
> > > * DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
> > > * Checked 152 log files, 16 edm output root files, 36 DQM output files
> > 
> > 
> > As expected, there are changes in variables related to TICL. I see also some changes in EgammaV/ElectronMcFakeValidator, ElectronMcSignalValidator, ElectronMcSignalValidatorPt1000. @rovere did you expect that?
> > https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_1_X_2020-11-17-1100+d355e5/39938/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44/EgammaV.html
> 
> Yes, I do expect them since #31766 (backport of #31527) by @SohamBhattacharya is feeding electrons with `TICL` products.

I confirm that Egamma related changes are expected as they are now seeded by TICL products, as Marco has pointed out.

jfernan2 commented:
+1

perrotta commented:
This PR provides backport of the following PRs merged in the master: #29837 (w/o Configuration/StandardSequences changes) #30007 #30815 #31013 (partial) #31031 #31730 #31881 #31906 #32021. While comparing with what merged in the HEAD of the master, I noticed the following points for which I'd like to get either an answer, or a fix.

- This PR implicitely also backports #30061 ("Added esConsumes to modules in RecoHGCal/TICL and related packages") for all RecoHGCal/TICL packages but not for HGCalEgammaIDHelper in RecoEgamma/EgammaTools: was that intentional? Why not to apply it also to HGCalEgammaIDHelper?
In any case, that PR needed a fix for Firework, #31536: I imagine the same fix will be needed in 10_6. Could you please check and, if needed, include also that fix in the backport?

- Please remove `#include "RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h"` from /RecoLocalCalo/HGCalRecAlgos/interface/HGCal3DClustering.h to avoid circular dependencies (see #31344)

- The differences wrt master observed in RecoLocalCalo/HGCalRecProducers/plugins/HGCalCLUEAlgo.cc show that PR #32022 ("Restore distance cut in CLUE position calculation") was not backported, while probably it also should

- Differences in RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc are due to #30806: since you are backporting quite a lot of stuffs, it wouldn't be bad to also backport this one, limitately to HGCalRecHitWorkerSimple.cc (i.e. just copy that file from the master)

- The `if` in /RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalMultiCluster.cc is better placed here than in the master: it wouldn't be bad to move it back also in the master, then

perrotta commented on RecoLocalCalo/HGCalRecProducers/python/hgcalLayerClusters_cff.py:
```diff
@@ -6,29 +6,37 @@
+        noiseMip = hgchebackDigitizer.digiCfg.noise.value()
+    )
+)
 
```
One of these two empty lines could be removed, both for the aesthetics and also to get rid of spurious differences with what in the master

rovere commented:
> This PR provides backport of the following PRs merged in the master: #29837 (w/o Configuration/StandardSequences changes) #30007 #30815 #31013 (partial) #31031 #31730 #31881 #31906 #32021. While comparing with what merged in the HEAD of the master, I noticed the following points for which I'd like to get either an answer, or a fix.
> 
> * This PR implicitely also backports #30061 ("Added esConsumes to modules in RecoHGCal/TICL and related packages") for all RecoHGCal/TICL packages but not for HGCalEgammaIDHelper in RecoEgamma/EgammaTools: was that intentional? Why not to apply it also to HGCalEgammaIDHelper?

I believe you meant #31013. If not, please explain a little better what you are asking for. PR #31013 has been almost fully back-ported. I skipped only the files that have been affected by #30061, that introduced a feature not-needed in `11_1` that would have caused massive editing and extra dependencies. Also, the packages you mention explicitly, namely `RecoEgamma/EgammaTools` have been fully updated to the `esConsumes`.

>   In any case, that PR needed a fix for Firework, #31536: I imagine the same fix will be needed in 10_6. Could you please check and, if needed, include also that fix in the backport?

No fix is needed because #30061, that introduced the bug, has not been back-ported on purpose.

> * Please remove `#include "RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h"` from /RecoLocalCalo/HGCalRecAlgos/interface/HGCal3DClustering.h to avoid circular dependencies (see #31344)

A plain backport would touch `TrackingTools/KalmanUpdators`, bringing in dependencies I'd prefer to avoid. I could manually decompose #31344 and pick whatever is related to `HGCAL` but this would be out-of-scope with this PR and will not get rid of the circular dependencies present in the other packages. I'd keep the content of this already big PR to the bare minimum, w/o #31344 or any segmentation of that.

> * The differences wrt master observed in RecoLocalCalo/HGCalRecProducers/plugins/HGCalCLUEAlgo.cc show that PR #32022 ("Restore distance cut in CLUE position calculation") was not backported, while probably it also should

I thought about that, as well. Given the little benefit that it brought to the reconstruction, I decided to leave that out (also because it came later wrt the original porting contained in this PR). Given the limited amount of changes, I could add that at the very bottom of this PR. This will trigger a full validation.

> * Differences in RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc are due to #30806: since you are backporting quite a lot of stuffs, it wouldn't be bad to also backport this one, limitately to HGCalRecHitWorkerSimple.cc (i.e. just copy that file from the master)

The same comment I made for PR #31344 apply here, too. This is not meant as a PR synching `11_1` to `HEAD`. The content of #30806 related to `HGCAL` are also minimal.

> * The `if` in /RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalMultiCluster.cc is better placed here than in the master: it wouldn't be bad to move it back also in the master, then

This I do not understand :sweat_smile:

perrotta commented:
> > This PR provides backport of the following PRs merged in the master: #29837 (w/o Configuration/StandardSequences changes) #30007 #30815 #31013 (partial) #31031 #31730 #31881 #31906 #32021. While comparing with what merged in the HEAD of the master, I noticed the following points for which I'd like to get either an answer, or a fix.
> > 
> > * This PR implicitely also backports #30061 ("Added esConsumes to modules in RecoHGCal/TICL and related packages") for all RecoHGCal/TICL packages but not for HGCalEgammaIDHelper in RecoEgamma/EgammaTools: was that intentional? Why not to apply it also to HGCalEgammaIDHelper?
> 
> I believe you meant #31013. If not, please explain a little better what you are asking for. PR #31013 has been almost fully back-ported. I skipped only the files that have been affected by #30061, that introduced a feature not-needed in `11_1` that would have caused massive editing and extra dependencies. Also, the packages you mention explicitly, namely `RecoEgamma/EgammaTools` have been fully updated to the `esConsumes`.
> 
> > In any case, that PR needed a fix for Firework, #31536: I imagine the same fix will be needed in 10_6. Could you please check and, if needed, include also that fix in the backport?
> 
> No fix is needed because #30061, that introduced the bug, has not been back-ported on purpose.
> 

Ok, great! I simply wanted to check

> > * Please remove `#include "RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h"` from /RecoLocalCalo/HGCalRecAlgos/interface/HGCal3DClustering.h to avoid circular dependencies (see #31344)
> 
> A plain backport would touch `TrackingTools/KalmanUpdators`, bringing in dependencies I'd prefer to avoid. I could manually decompose #31344 and pick whatever is related to `HGCAL` but this would be out-of-scope with this PR and will not get rid of the circular dependencies present in the other packages. I'd keep the content of this already big PR to the bare minimum, w/o #31344 or any segmentation of that.
> 

Well, I asked to remove that include from HGCal3DClustering.h since it does nothing. Of course that is a minor point, and if there are no other changes needed in the code I won't insist for applying only that innocent fix,

> > * The differences wrt master observed in RecoLocalCalo/HGCalRecProducers/plugins/HGCalCLUEAlgo.cc show that PR #32022 ("Restore distance cut in CLUE position calculation") was not backported, while probably it also should
> 
> I thought about that, as well. Given the little benefit that it brought to the reconstruction, I decided to leave that out (also because it came later wrt the original porting contained in this PR). Given the limited amount of changes, I could add that at the very bottom of this PR. This will trigger a full validation.
> 

Ok, this is that you (the HGCal/TICL group) has to decide. Since we decided that the TICL reconstruction in 11_1 can be updated without being constrained by the "no change" rule for the frozen production releases, I would personally profit and backport the best possible version of it, But if you are convinced that what you decided to backport is enough for the studies that you want to perform with 11_1, it won't be me to insist for completing the backport

> > * Differences in RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc are due to #30806: since you are backporting quite a lot of stuffs, it wouldn't be bad to also backport this one, limitately to HGCalRecHitWorkerSimple.cc (i.e. just copy that file from the master)
> 
> The same comment I made for PR #31344 apply here, too. This is not meant as a PR synching `11_1` to `HEAD`. The content of #30806 related to `HGCAL` are also minimal.
> 

I agree. As before, since you are doing it syncing 11_1 to HEAD when it is costless may bring the advantage of simplifying future possible backports. On the other hand you probably won't need further TICL related backports to 11_1, and therefore I also won't inisist on it

> > * The `if` in /RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalMultiCluster.cc is better placed here than in the master: it wouldn't be bad to move it back also in the master, then
> 
> This I do not understand 😅

I simply meant that here, in 11_X, you "skip empty multiclusters" before "filter using trackster PID", while in the master you follow the opposit order. And the order chosen for this "backport" looks like being more convenient performance wise than the one in the master. My suggestion was to make a simple a trivial PR in the master that reverts the order to be the same as in this PR

perrotta commented:
Trying to summarize the exchange we had yesterday in this thread: I raised a few concerns about the backport originated by the comparison of the backported code and the one in the master, and @rovere satisfactorily answered all them. At this point basically we (or, better, the TICL group) have to decide whether we want to backport in 11_1 also the fix integrated in the master with #32022 or not. The other fixes I suggested can be considered at the level of cosmetics, or slighlty more.

Therefore, there are three possibilities in front:
1) You decide to backport also #32022, and we do it in this PR, which has to be modified accordingly. While doing so one can decide whether to update also with the other items I pointed out (1a) or not (1b)
2) You decide to backport also #32022, and we do it in a separate backport PR to be submitted as soon as this one is merged in 11_1. I will sign this PR as it is now, in that case. While preparing the new backport PR one can decide whether to update also with the other items I pointed out (2a) or not (2b)
3) You decide not to backport #32022, even if it was already verified to be a (small) improvement for TICL. I will sign this PR as it is now, in that case.

I would personally opt for either (1) or (2), but this is a decision to be based upon physics judgement, and taking into account the needs for the productions to be prepared with the future new 11_1 release: therefore, I rather let the groups involved to take a decision.

rovere committed:
Backport(partcial) of #31344

rovere committed:
Backport(partial) of #30806

cmsbuild commented:
Pull request #31907 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

rovere commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
+1
Tested at: de99de0fce59474c81e020db89576f6d898189de
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10893/summary.html
CMSSW: CMSSW_11_1_X_2020-11-20-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f9b338/10893/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3943 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2784828
 - DQMHistoTests: Total failures: 13638
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2771140
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 28 edm output root files, 36 DQM output files

perrotta commented:
+1
- Backport of #29837,  #30007 #30815 #31013 (partial) #31031 #31730 #31881 #31906 #32021 #32022, plus a few additional TICL related improvements/cleanings already merged in the master; in most cases the TICL code is now aligned to the one currently in the CMSSW master 
- Differences in Phase2 outputs wrt the baseline are as expected due to the backported PRs. It was decided at the ORP that the "no change" rule for closed releases can be relaxed now for 11_1_X, trying to have the most updated reconstruction for the forthcoming Phase2 HLT TDR studies
- New objects saved (ticlTrackstersMerge, ticlTrackstersTrkEM, ticlMultiClustersFromTrackstersTrkEM, ticlTrackstersHFNoseEM, ticlTrackstersHFNoseMIP, ticlTrackstersHFNoseMerge), others removed (ticlCandidateFromTracksters)

jfernan2 commented:
+1

kpedro88 commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_11_1_X IBs (tests are also fine) and once validation in the development release cycle CMSSW_11_2_X is complete. This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1

hatakeyamak commented:
@rovere I see that this is now integrated to 11_1.
Does this run with injectTICLintoPF without ChargedHadronPFTrackIsolationProducer fix (backport of #32202) or do we need this backport?
